### PR TITLE
CR for Job #21495 - Add limit to file size imported into asset server

### PR DESCRIPTION
**Pull request by mnafees**
Code Review for Job #21495 - Workitem available at https://www.worklist.net/21495

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3462/HighFidelity-Beta-PR11412-915769b9d813bbc960ff78527ba9653eb3b3c6d0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3462/HighFidelity-Beta-PR11412-915769b9d813bbc960ff78527ba9653eb3b3c6d0.exe

**Comment by mnafees** [comment]
## Test Plan

1. Start domain-server and assignment-client and make sure asset-server is running
2. Open http://local host:40100/settings and navigate to the "Asset Server" section in the Advanced menu section.
3. Change the file size limit to any desired size in MBytes
4. Click on the "Save and Restart" button to restart the assignment-client 
5. Open Interface and go to Edit -> Asset Browser 
6. Try to upload a file larger than the file limit to see the error message.

You can go back and change the size to 0 in settings which means a file of any size can be uploaded to the Asset Server.

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

Start domain-server and assignment-client and make sure asset-server is running **Check**
Open http://localhost:40100/settings and navigate to the "Asset Server" section in the Advanced menu section. **Check**
Change the file size limit to any desired size in MBytes **Check, Tested 1MB Cap**
Click on the "Save and Restart" button to restart the assignment-client **Check**
Open Interface and go to Edit -> Asset Browser **Check**
Try to upload a file larger than the file limit to see the error message. **Check, error said file too large when using over size limit file, when under worked as normal.**
You can go back and change the size to 0 in settings which means a file of any size can be uploaded to the Asset Server. **Check, now uploading larger file works as cap is removed successfully.**

**Pass**

File: assignment-client/src/assets/AssetServer.cpp [hunk 1]
```diff
@@ -74,7 +74,8 @@ void updateConsumedCores() {
 
 AssetServer::AssetServer(ReceivedMessage& message) :
     ThreadedAssignment(message),
-    _taskPool(this)
+    _taskPool(this),
+    _filesizeLimit(0)
```
Comment (Atlante45) [inline assignment-client/src/assets/AssetServer.cpp]: Should be initialized in the header.
And it should be initialized to `MAX_UPLOAD_SIZE`

File: assignment-client/src/assets/AssetServer.cpp [hunk 2]
```diff
@@ -199,6 +200,19 @@ void AssetServer::completeSetup() {
         setFinished(true);
     }
 
+    // get file size limit for an asset
+    static const QString ASSETS_FILESIZE_LIMIT_OPTION = "assets_filesize_limit";
+    auto assetsFilesizeLimitJSONValue = assetServerObject[ASSETS_FILESIZE_LIMIT_OPTION];
+    bool ok = false;
+    auto assetsFilesizeLimit = assetsFilesizeLimitJSONValue.toVariant().toInt(&ok);
```
Comment (Atlante45) [inline assignment-client/src/assets/AssetServer.cpp]: QJsonValue can be cast directly:
```cpp
auto assetsFilesizeLimit = assetsFilesizeLimitJSONValue.toInt(MAX_UPLOAD_SIZE);
```

Let's also not finish if the value can't be read.

File: assignment-client/src/assets/AssetServer.cpp [hunk 3]
```diff
@@ -199,6 +200,19 @@ void AssetServer::completeSetup() {
...
+
+    if (!ok) {
+        qCritical() << "Received an assets max file size from the domain-server that could not be parsed. Stopping assignment.";
+        setFinished(true);
+        return;
+    }
+
+    _filesizeLimit = assetsFilesizeLimit * 1000 * 1000; // store bytes
```
Comment (Atlante45) [inline assignment-client/src/assets/AssetServer.cpp]: Magic number: Explicitly name `1000 * 1000` (ex: `BYTES_PER_MEGABYTES`)
Pretty sure it's already defined in the code somewhere too.

File: assignment-client/src/assets/UploadAssetTask.cpp [hunk 4]
```diff
@@ -50,6 +51,8 @@ void UploadAssetTask::run() {
     
     if (fileSize > MAX_UPLOAD_SIZE) {
         replyPacket->writePrimitive(AssetServerError::AssetTooLarge);
+    } else if (_filesizeLimit != 0 && fileSize > _filesizeLimit) {
+        replyPacket->writePrimitive(AssetServerError::FilesizeLimitCrossed);
```
Comment (Atlante45) [inline assignment-client/src/assets/UploadAssetTask.cpp]: There's a hardcoded limit and an adjustable limit here.
Remove the hardcoded limit and default the adjustable one to that hard-coded value.

File: libraries/networking/src/AssetUpload.h [hunk 5]
```diff
@@ -33,7 +33,8 @@ class AssetUpload : public QObject {
         TooLarge,
         PermissionDenied,
         FileOpenError,
-        ServerFileError
+        ServerFileError,
+        FilesizeLimitCrossed
```
Comment (Atlante45) [inline libraries/networking/src/AssetUpload.h]: There is already a `TooLarge` error.

File: libraries/networking/src/AssetUtils.h [hunk 6]
```diff
@@ -41,7 +41,8 @@ enum AssetServerError : uint8_t {
     AssetTooLarge,
     PermissionDenied,
     MappingOperationFailed,
-    FileOperationFailed
+    FileOperationFailed,
+    FilesizeLimitCrossed
```
Comment (Atlante45) [inline libraries/networking/src/AssetUtils.h]: Same here, there is `AssetTooLarge`

**Comment by mnafees** [comment]
Making the required changes and pushing the diff

**Comment by themelissabrown** [comment]
@mnafees please let me know when this is ready for review again

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3952/HighFidelity-Beta-PR11412-915769b9d813bbc960ff78527ba9653eb3b3c6d0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3952/HighFidelity-Beta-PR11412-915769b9d813bbc960ff78527ba9653eb3b3c6d0.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3998/HighFidelity-Beta-PR11412-0764a0bb2bc436fc2b116bef81c06243916d2e72.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4010/HighFidelity-Beta-PR11412-05998927c0a49f837364f2441669acc199ae281e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4010/HighFidelity-Beta-PR11412-05998927c0a49f837364f2441669acc199ae281e.exe

**Comment by Atlante45** [comment]
@mnafees the build fail is because of a new warning introduced by this PR:

> AssetServer.cpp:409, GNU C Compiler 4 (gcc), Priority: Normal
> comparison between signed and unsigned integer expressions [-Wsign-compare]

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4072/HighFidelity-Beta-PR11412-52e29c07cc9c06b3cfe43940ebf5bc31ccff1a39.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4072/HighFidelity-Beta-PR11412-52e29c07cc9c06b3cfe43940ebf5bc31ccff1a39.exe

**Pull request merged**